typing/ctype.ml

```diff
@@ -263,6 +263,36 @@ let newconstr path tyl = newty (Tconstr (path, tyl, ref Mnil))
 
 let none = newty (Ttuple [])                (* Clearly ill-formed type *)
 
+(**** information for [Typecore.unify_pat_*] ****)
+
+module Pattern_env : sig
+  type hidden_env
+  type t = private
+    { mutable env : hidden_env;
+      equations_scope : int;
+      allow_recursive_equations : bool; }
+  val make: Env.t -> equations_scope:int -> allow_recursive_equations:bool -> t
+  val copy: ?equations_scope:int -> t -> t
+  val get_env: t -> Env.t
+  val set_env: t -> Env.t -> unit
```

Comment:
The module is much simpler now, which is good. But then I am not sure that I see the point of making `hidden_env` an abstract type: why not just let consumers access and mutate the `env` field, instead of providing `{get,set}_env` functions to do it? The important thing is to have an explicit `copy` operation I think.

Reply:
The intention was to clearly mark accesses to env and make future changes / refactoring easier.  I did not find much drawback doing this.

Reply:
I think that it is unnecessary / makes the code heavier than it needs to be, but you are the boss of your own code and I am okay with the proposed version. Let's merge when the CI agrees.

Reply:
Note that I tend to agree with @gasche, that getter/setter à la Java looks a bit like a premature abstraction. But that's not a problematic abstraction to maintain (or remove in the future), so this sounds fine to me.

Reply:
I am now getting persuaded to remove accessors. Let me show you the version without them before merging.

Reply:
Looks good, thanks!